pkg/manager/member/tidb_init_manager.go

```diff
@@ -0,0 +1,428 @@
+	"github.com/pingcap/tidb-operator/pkg/label"
+)
+
+const (
+	initContainerName   = "wait"
+	containerName       = "mysql-client"
+	passwdKey           = "password"
+	passwdPath          = "/etc/tidb/password"
+	sqlKey              = "init-sql"
+	sqlPath             = "init.sql"
+	sqlDir              = "/data"
+	startScriptPath     = "start_script.py"
+	initStartScriptPath = "init_start_script.sh"
+	startScriptDir      = "/usr/local/bin"
+	startKey            = "start-script"
+	initStartKey        = "init-start-script"
+	// initContainerImage is the image for init container
+	initContainerImage = "busybox:1.31.1"
+)
+
+// InitManager implements the logic for syncing TidbInitializer.
+type InitManager interface {
+	// Sync	implements the logic for syncing TidbInitializer.
+	Sync(*v1alpha1.TidbInitializer) error
+}
+
+type tidbInitManager struct {
+	jobLister    batchlisters.JobLister
+	cli          versioned.Interface
+	tiLister     listers.TidbInitializerLister
+	typedControl controller.TypedControlInterface
+}
+
+// NewTiDBInitManager return tidbInitManager
+func NewTiDBInitManager(
+	jobLister batchlisters.JobLister,
+	cli versioned.Interface,
+	tiLister listers.TidbInitializerLister,
+	typedControl controller.TypedControlInterface,
+) InitManager {
+	return &tidbInitManager{
+		jobLister,
+		cli,
+		tiLister,
+		typedControl,
+	}
+}
+
+func (tm *tidbInitManager) Sync(ti *v1alpha1.TidbInitializer) error {
+	name := controller.TiDBInitializerMemberName(ti.Spec.Clusters.Name)
+	ns := ti.Namespace
+	cm := &corev1.ConfigMap{}
+
+	exist, err := tm.typedControl.Exist(client.ObjectKey{
+		Namespace: ns,
+		Name:      name,
+	}, cm)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		err = tm.syncTiDBInitConfigMap(ti)
+		if err != nil {
+			return err
+		}
+	}
+
+	job, err := tm.jobLister.Jobs(ns).Get(name)
+	if err == nil {
+		// job has been created
+		return tm.updateStatus(ti.DeepCopy(), job)
+	}
+
+	if !errors.IsNotFound(err) {
+		return fmt.Errorf("TiDBInitializer %s/%s get job %s failed, err: %v", ns, ti.Name, name, err)
+	}
+
+	return tm.syncTiDBInitJob(ti)
+}
+
+func (tm *tidbInitManager) updateStatus(ti *v1alpha1.TidbInitializer, job *batchv1.Job) error {
+	if job == nil || ti == nil {
+		return nil
+	}
+	phase := v1alpha1.InitializePhaseRunning
+	if len(job.Status.Conditions) > 0 {
+		for _, c := range job.Status.Conditions {
+			if c.Type == batchv1.JobComplete && c.Status == corev1.ConditionTrue {
+				phase = v1alpha1.InitializePhaseCompleted
+				break
+			}
+			if c.Type == batchv1.JobFailed && c.Status == corev1.ConditionTrue {
+				phase = v1alpha1.InitializePhaseFailed
+				break
+			}
+		}
+	}
+	var update bool
+	if !apiequality.Semantic.DeepEqual(ti.Status.JobStatus, job.Status) {
+		job.Status.DeepCopyInto(&ti.Status.JobStatus)
+		update = true
+	}
+	if ti.Status.Phase != phase {
+		ti.Status.Phase = phase
+		update = true
+	}
+	if update {
+		return tm.retryUpdateStatus(ti)
+	}
+	return nil
```

Comment:
This could be simplified using `GuaranteedUpdate`, we don't have to check the equality in client-side actually, if the update do not mutate the object, the `resourceVersion` of object won't be changed, that is, the object will not be affected at all.

Reply:
Check before creation can reduce the API call to kube-apiserver, and I have updated to use `GuaranteedUpdate`.